rts/motoko-rts/src/bitrel.rs

```diff
@@ -0,0 +1,107 @@
+//! This module implements a simple subtype cache used by the compiler (in generated code)
+
+use crate::constants::WORD_SIZE;
+use crate::idl_trap_with;
+use crate::mem_utils::memzero;
+use crate::types::Words;
+
+const BITS: u32 = 2;
+
+#[repr(packed)]
+pub struct BitRel {
+    /// Pointer into the bit set
+    pub ptr: *mut u32,
+    /// Pointer to the end of the bit set
+    /// must allow at least 2 * size1 * size2 bits
+    pub end: *mut u32,
+    pub size1: u32,
+    pub size2: u32,
+}
+
+impl BitRel {
+    pub(crate) fn words(size1: u32, size2: u32) -> u32 {
+        return ((2 * size1 * size2 * BITS) + (usize::BITS - 1)) / usize::BITS;
+    }
+
+    pub(crate) unsafe fn init(&self) {
+        if (self.end as usize) < (self.ptr as usize) {
+            idl_trap_with("BitRel invalid fields");
+        };
+
+        let bytes = ((self.end as usize) - (self.ptr as usize)) as u32;
+        if (2 * self.size1 * self.size2 * BITS) > bytes * 8 {
+            idl_trap_with("BitRel not enough bytes");
+        };
+        memzero(self.ptr as usize, Words(bytes / WORD_SIZE));
+    }
```

Comment:
Not sure whether some bits may remain uninitialized, if bytes is not a multiple of word size.

Reply:
Good catch - will need to check what this was doing - wrote it in April....

Reply:
Note to self: I think this is ok because 
```
(self.end as usize) - (self.ptr as usize) = words(self.size1,self.size2)*WORD_SIZE
```

by the way this is initialized in `compile.ml` (so bytes is a mutlple of WORD_SIZE) but I should add that as an assert or check to make the code clearer.

Reply:
I changed the code to make it clear that bytes muse be a multiple of WORD_SIZE (by adjusting the previous check)